lib/charms/tls_certificates_interface/v4/tls_certificates.py

```diff
@@ -1212,3 +1228,237 @@ def _get_csr_secret_label(self, csr: CertificateSigningRequest) -> str:
 
     def _get_unit_number(self) -> str:
         return self.model.unit.name.split("/")[1]
+
+
+class TLSCertificatesProvidesV4(Object):
+    """TLS certificates provider class to be instantiated by TLS certificates providers."""
+
+    def __init__(self, charm: CharmBase, relationship_name: str):
+        super().__init__(charm, relationship_name)
+        self.framework.observe(charm.on[relationship_name].relation_changed, self._configure)
+        self.framework.observe(charm.on.update_status, self._configure)
+        self.charm = charm
+        self.relationship_name = relationship_name
+
+    def _configure(self, _: EventBase) -> None:
+        """Handle update status and tls relation changed events.
+
+        This is a common hook triggered on a regular basis.
+
+        Revoke certificates for which no csr exists
+        """
+        if not self.model.unit.is_leader():
+            return
+        self._revoke_certificates_for_which_no_csr_exists()
+
+    def _revoke_certificates_for_which_no_csr_exists(self) -> None:
+        provider_certificates = self._get_provider_certificates()
+        requirer_csrs = self.get_certificate_requests()
+        for provider_certificate in provider_certificates:
+            for requirer_csr in requirer_csrs:
+                if (
+                    provider_certificate.certificate_signing_request
+                    == requirer_csr.certificate_signing_request
+                ):
+                    tls_relation = self._get_tls_relations(relation_id=requirer_csr.relation_id)
+                    self._remove_provider_certificate(
+                        certificate=provider_certificate.certificate,
+                        relation=tls_relation[0],
+                    )
```

Comment:
If you are, then I am also reading it wrong. In any case, this could be simplified with something like:
```suggestion
        for provider_certificate in provider_certificates:
            if provider_certificate.certificate_signing_request not in requirer_csrs:
                tls_relation = self._get_tls_relations(relation_id=requirer_csr.relation_id)
                self._remove_provider_certificate(
                    certificate=provider_certificate.certificate,
                    relation=tls_relation[0],
                )
```

Reply:
Oops, indeed I had a mistake. The recommendation does not work though, `requirer_csr.relation_id` can't be called (requirer_csr is not defined)

Reply:
This should be fixed now